client/service/src/config.rs

```diff
@@ -191,3 +195,52 @@ impl Default for RpcMethods {
 		RpcMethods::Auto
 	}
 }
+
+/// The base path that is used for everything that needs to be write on disk to run a node.
+pub enum BasePath {
+	/// A temporary directory is used as base path and will be deleted when dropped.
+	#[cfg(not(target_os = "unknown"))]
+	Temporary(TempDir),
+	/// An path existing path on the disk.
+	Permanenent(PathBuf),
+}
+
+impl BasePath {
+	/// Create a temporary directory prefixed with "substrate" and use it as base path.
+	#[cfg(not(target_os = "unknown"))]
+	pub fn new_temp_dir() -> io::Result<BasePath> {
+		Ok(BasePath::Temporary(
+			tempfile::Builder::new().prefix("substrate").tempdir()?,
+		))
+	}
+
+	/// Create a base path based on an existing path on disk.
+	pub fn new<P: AsRef<Path>>(path: P) -> BasePath {
+		BasePath::Permanenent(path.as_ref().to_path_buf())
+	}
+
+	/// Create a base path from values describing the project.
+	#[cfg(not(target_os = "unknown"))]
+	pub fn from_project(qualifier: &str, organization: &str, application: &str) -> BasePath {
+		BasePath::new(
+			directories::ProjectDirs::from(qualifier, organization, application)
+				.expect("app directories exist on all supported platforms; qed")
+				.data_local_dir(),
+		)
+	}
+
+	/// Retrieve the base path.
+	pub fn path(&self) -> &Path {
```

Comment:
You want `Deref<Target = Path>`

Reply:
Not me! :grin: `TempDir` wasn't doing that and I'm not too fan of `Deref` myself so I used the same API than TempDir https://docs.rs/tempfile/3.1.0/tempfile/struct.TempDir.html#method.path

I don't mind implementing the `Deref` if you want it to but I would prefer implementing the `AsRef<Path>` like they do in TempDir.

Reply:
Why are you not a "fan" of `Deref`? What is the problem?

Reply:
First: because it is not as explicit as calling a method that will give you the underlying object.

For example if I think the variable `base_path` here is a PathBuf or Path I might be tempted dropping it, not realizing that this will trigger some very specific behavior (which is to delete the directory!):

```rust
{
    let base_path = ...
}

# ok I'm done with this path, continue running the Service
```

If you are not careful enough, you will be surprised.

Obviously for us right now it's easy because we saw it from the beginning but someone who don't know what happened to the base_path might think it is the base_path it used to be (a simple PathBuf).

Second: deleting a directory of a file that is being used is authorized in Linux so you don't even realize right away what you are doing. For example, you pointed out that the directory was deleted because of how the `run_service_until_exit` works and I sincerely didn't catch it. I even wrote a test and it wasn't visible. I put a message on the destructor, pressed CTRL+C, and I saw that the message was displayed exactly when I thought it would be so I believed it was working correctly.

Two reasons for me to not make this thing too smart.